### PR TITLE
Adding KnownKey

### PR DESCRIPTION
**Pull request by Licenser**
Adding a feature for accessing known keys faster.

### CODE REVIEW

**Comment by sunnygleason** [comment]
@Licenser this looks interesting -- does it just impact serde usage, because it's serialization/deserialization of a type with a more performant trait?

**Comment by Licenser** [comment]
It doesn't affect serde compatibility or serialisation no. The effect is for BorrowedValue.get() (and I got to add OwnedValue.get() (performance improvement when using KnownKey). And deserialisation (performance decrease when using the `known-key` feature)

File: src/known_key.rs [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+        }))
+    }
+}
+#[cfg(test)]
+mod tests {
+    #![allow(clippy::unnecessary_operation, clippy::non_ascii_literal)]
+    use super::*;
+    use crate::borrowed::*;
+    use crate::ValueTrait;
+
```
Comment (sunnygleason) [inline src/known_key.rs]: Maybe an end-to-end test somewhere (perhaps with a doc comment with usage somewhere in the API?) would make things a bit clearer? These tests avoid the JSON processing APIs (targeted & good for a unit test, my brain was looking for an illustrative integration test(s)? I dunno...)

**Comment by sunnygleason** [comment]
This seems very neat!

* I was looking for an illustration doc or integration test to "demo" usage
* could we add a benchmark case to show it works?

If you want to add a TODO comment/issue to the benchmarks in the right place, I'd be happy to pick that up!

**Comment by Licenser** [comment]
codecov also seems to be broken :/

File: src/known_key.rs [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+    ///
+    /// ```rust
+    /// use simd_json::*;
+    /// let object: BorrowedValue = json!({
+    ///   "answer": 42,
+    ///   "key": 7
+    /// }).into();
+    /// let known_key = KnownKey::from("answer");
+    /// assert_eq!(known_key.lookup(&object).unwrap(), &42);
+    /// ```
```
Comment (sunnygleason) [inline src/known_key.rs]: this is awesome!

**Comment by sunnygleason** [comment]
@Licenser this is very neat. So it's like precompilation of hash values on open objects (hashes) with known keys. Thank you for the implementation, and especially the docs, I think those turned out very nicely indeed!

File: src/known_key.rs (hunk 1 above)
Comment (Licenser) [reply src/known_key.rs]: added!

**Comment by Licenser** [comment]
Thanks! It is a bit of a specialised case you really have the need for to use it but it can give some significant performance boosts. At the app I'm working on in $job there were some benchmarks where this went from 140 MB/s to 150MB/s throughput on this which is a nice speedup :D

**Comment by sunnygleason** [comment]
> Thanks! It is a bit of a specialised case you really have the need for to use it but it can give some significant performance boosts. At the app I'm working on in $job there were some benchmarks where this went from 140 MB/s to 150MB/s throughput on this which is a nice speedup :D

That's interesting, I wonder if the field names add up to a lot of wasted copying as well. Could this case be further optimized to use `string -> int` indexes into a fixed-size `vec`, omitting fields?

**Comment by Licenser** [comment]
Nah it's because we've build a language to do ETL on data, which happens to be JSON(esque) so JSONPath like expressions `this.that.here.there` do heavily rely on lookups in Objects. But since the script is only compiled ones but executed multiple times pre-paying for hashing and memorising it makes a difference.

**Comment by Licenser** [comment]
released as v0.1.25 :)

**Comment by sunnygleason** [comment]
sounds delightful! 😀

**Pull request merged**